components/script/dom/document.rs

```diff
@@ -722,9 +726,12 @@ impl Document {
                     if document.page_showing.get() {
                         return;
                     }
-                    // Step 4.6.2
+                    // Step 4.6.2 Set document's page showing flag to true.
```

Comment:
We are here at https://html.spec.whatwg.org/multipage/#reactivate-a-document, but it's not clear if we are calling this in the right place. `#reactivate-a-document` is called from https://html.spec.whatwg.org/multipage/browsing-the-web.html#update-document-for-history-step-application, which seems to be rather at https://github.com/servo/servo/blob/ad01342f00089cbddb252d54ed55f90a20ce43da/components/script/dom/history.rs#L86

It would be good to figure out more about how Servo's implementation maps to the spec, and either move your changes to other parts, or highlight problems to be resolved as follow-ups. 

To me, the `set_activity` handler, where we are here, seems to correspond more to https://html.spec.whatwg.org/multipage/interaction.html#page-visibility:update-the-visibility-state (since we are receiving a message from the constellation telling us the system visibility has changed), but the code here seems to be #reactivate-a-document.

Answering this question requires looking at what messages the constellation sends for history updates, and for system visibility(perhaps both go through here, but we should at least clarify it).

Reply:
I tried my best to track all changes, but I think https://html.spec.whatwg.org/multipage/#history-traversal has changed so much that it will require a big overhaul. That means not only [#reactivate-a-document](https://html.spec.whatwg.org/multipage/browsing-the-web.html#reactivate-a-document) but also [#make-active](https://html.spec.whatwg.org/multipage/browsing-the-web.html#make-active) needs to be added. But many of their callers do not really exist yet in Servo, which brings us back to the problem of big overhaul again.

The good news is at least `set_activity` can be called correctly by pipeline and constellation when creating new context. So, the visibility state would be visible when the document is created and the history session is changed. Maybe we should add a follow up issue to track how to refactor document and history to match latest ["Populating a session history entry"](https://html.spec.whatwg.org/multipage/browsing-the-web.html#populating-a-session-history-entry) and ["Updating the document"](https://html.spec.whatwg.org/multipage/browsing-the-web.html#updating-the-document)?

Reply:
Yes opening an issue to document these problems is a good idea, also I think we should update the spec link above to be just [#make-active](https://html.spec.whatwg.org/multipage/browsing-the-web.html#make-active) (since history traversal calls into this, but it's not clear that we are really running these steps as part of history traversal).

Reply:
I opened #32687 and added a TODO comment to indicate future updates.